importers/link_16_importer.py

```diff
@@ -0,0 +1,199 @@
+import os
+from datetime import datetime, timedelta
+
+from tqdm import tqdm
+
+from pepys_import.core.formats import unit_registry
+from pepys_import.core.formats.location import Location
+from pepys_import.core.validators import constants
+from pepys_import.file.highlighter.support.combine import combine_tokens
+from pepys_import.file.importer import CANCEL_IMPORT, Importer
+from pepys_import.utils.sqlalchemy_utils import get_lowest_privacy
+from pepys_import.utils.text_formatting_utils import (
+    custom_print_formatted_text,
+    format_error_message,
+)
+from pepys_import.utils.unit_utils import convert_absolute_angle, convert_distance, convert_speed
+
+V1_HEADER = "PPLI"
+V2_HEADER = "Xmt/Rcv"
+
+
+class Link16Importer(Importer):
+    """Importer to handle two different formats of track information that are
+    transmitted using Link-16 encoding
+    """
+
+    def __init__(self):
+        super().__init__(
+            name="Link-16 Format Importer",
+            validation_level=constants.BASIC_LEVEL,
+            short_name="Link-16 Importer",
+            default_privacy="Private",
+            datafile_type="Link-16",
+        )
+        self.version = 1
+
+    def can_load_this_type(self, suffix):
+        return suffix.upper() == ".CSV"
+
+    def can_load_this_filename(self, filename):
+        return True
+
+    def can_load_this_header(self, header):
+        # V1 starts w/ PPLI
+        # V2 starts w/ Xmt/Rcv
+        return header.startswith(V1_HEADER) or header.startswith(V2_HEADER)
+
+    def can_load_this_file(self, file_contents):
+        return True
+
+    def _load_this_file(self, data_store, path, file_object, datafile, change_id):
+        # Read the base timestamp from the filename
+        filename, ext = os.path.splitext(os.path.basename(path))
+        datetime_string = self.extract_timestamp(filename)
+        self.base_timestamp = self.timestamp_to_datetime(datetime_string)
+        self.current_hour = 0
```

Comment:
@mew-nsc - I _think_ we need to put the hours from the filename into this field.  Sample data showed that sometimes there was a delay of up to 10 minutes between the time in the filename and the first reading.

If the filename is `09-58-16`, and the first reading is `03:12.23`, then we need to increment the hours to generate a time of `10:03:12`.

Reply:
It would probably be good to clarify the behaviour of this - the existing parsers take a value for time in rather than reading it from the filename. So I'd taken the offset by the start time from there - so for the case you've given, the result would be 9:58:16 + 00:03:12.23 = 10:01:28

Running the file through the REP parsers with those values (as `start_time` = 095816) gives an output in the REP file of 10:01:51 - Not quite the number I was expecting but delving further it looks to be caused by a bug in the parser code around handling of the milliseconds.

Reply:
Aah,  so that's why it's a duration in the code.  I'm pleased we've explored it further.

From looking at the data it's not a duration. It's time.   So, when the time of the file is 12-36-32, the first line will be something like 37:12.01. It's a minute/seconds equal to, or shortly after the time value in the filename.